torch/csrc/autograd/python_engine.cpp

```diff
@@ -70,6 +70,16 @@ variable_list PythonEngine::execute(
   }
 }
 
+variable_list PythonEngine::execute_with_graph_task(
+    std::shared_ptr<GraphTask> graph_task,
+    std::shared_ptr<Node> graph_root) {
+  try {
+    return Engine::execute_with_graph_task(graph_task, graph_root);
+  } catch (python_error& e) {
+    e.restore();
```

Comment:
This looks pretty weird to me. What it kind of looks like is that you reraised a `python_error` from another thread but forgot to restore the Python error context at the time you threw the error, so now you have to "fix" it here. Wouldn't it be better to just handle it properly at the rethrowing site?

Reply:
There is a case where this method is called directly from `DistEngine` in that case, we would like to restore the error. I'll change this to only do `e.restore` if the error is not set already.